wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,259 @@
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn bound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &BoundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(state, &Decl::Bound(decl.clone()), &self.exceptable_nodes());
+        }
+    }
+
+    fn unbound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &UnboundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(
+                state,
+                &Decl::Unbound(decl.clone()),
+                &self.exceptable_nodes(),
+            );
+        }
+    }
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Declaration names should not include their type"
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Declaration names should not include their type as a prefix or suffix. \
+        This makes the code more verbose and often redundant. For example, use \
+        'input_file' instead of 'file_input' or 'myFile'."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::OutputSectionNode,
+        ])
+    }
+}
+
+/// Check declaration name for type prefixes or suffixes
+fn check_decl_name(
+    state: &mut Diagnostics,
+    decl: &Decl,
+    exceptable_nodes: &Option<&'static [SyntaxKind]>,
+) {
+    let name = decl.name();
+    let name_str = name.as_str();
+    let name_lower = name_str.to_lowercase();
+    
+    // Get the declaration type
+    let ty = decl.ty();
+    
+    // Extract base type name(s) using AST methods
+    let mut type_names = Vec::new();
+    
+    // Add the full type name
+    type_names.push(ty.to_string());
+    
+    // If it's an array, add the element type
+    if let Some(array_type) = ty.as_array_type() {
+        // Add the element type 
+        type_names.push(array_type.element_type().to_string());
+    }
+    
+    // If it's a map, add the key and value types
+    if let Some(map_type) = ty.as_map_type() {
+        let (key, value) = map_type.types();
+        type_names.push(key.to_string());
+        type_names.push(value.to_string());
+    }
+    
+    // If it's a pair, add the left and right types
+    if let Some(pair_type) = ty.as_pair_type() {
+        let (left, right) = pair_type.types();
+        type_names.push(left.to_string());
+        type_names.push(right.to_string());
+    }
```

Comment:
This won't work as expected for anything complexly nested. e.g. `Array[Pair[Int, Map[String, File]]]`